keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1127 @@
+container separately at container creation time with the `CreateContainer`
+request.
+
+```diff
+ message PodSandboxConfig {
+ 
+ ...
+ 
+     // Optional configurations specific to Linux hosts.
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++
++    // Kubernetes resource spec of the containers in the pod.
++    PodResourceConfig pod_resources = 10;
+ }
+ 
++// PodResourceConfig contains information of all resources requirements of
++// the containers of a pod.
++message PodResourceConfig {
++    repeated ContainerResourceConfig containers = 1;
++}
+ 
++// ContainerResourceConfig contains information of all resource requirements of
++// one container.
++message ContainerResourceConfig {
++    // Name of the container
++    string name= 1;
++
++    // Type of the container
++    ContainerType type= 2;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 3;
++
++    // Mounts for the container.
++    repeated Mount mounts = 4;
++
++    // Devices for the container.
++    repeated Device devices = 5;
++
++    // CDI devices for the container.
++    repeated CDIDevice CDI_devices = 6;
++}
+
++enum ContainerType {
++    INIT_CONTAINER    = 0;
++    SIDECAR_CONTAINER = 1;
++    REGULAR_CONTAINER = 2;
++}
+```
+
+#### CreateContainer
+
+The ContainerConfig message (used in CreateContainer request) is extended to
+contain unmodified resource requests from the PodSpec.
+
+```diff
++import "k8s.io/apimachinery/pkg/api/resource/generated.proto";
+
+ message ContainerConfig {
+ 
+ ...
+ 
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
+ 
+     // CDI devices for the container.
+     repeated CDIDevice CDI_devices = 17;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 18;
+ }
+ 
++// KubernetesResources contains the resource requests and limits as specified
++// in the Kubernetes core API ResourceRequirements.
++message KubernetesResources {
++    // Requests and limits from the Kubernetes container config.
++    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> requests = 1;
++    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> limits = 2;
++}
+```
+
+Note that mounts, devices, CDI devices are part of the ContainerConfig message
+but are left out of the diff snippet above.
+
+The resources (mounts, devices, CDI devices, Kubernetes resources) in the
+CreateContainer request should be identical to what was (pre-)informed in the
+RunPodSandbox request. If they are different, the CRI runtime may fail the
+container creation, for example because changes cannot be applied after a
+VM-based Pod has been created.
+
+#### UpdateContainerResourcesRequest
+
+The UpdateContainerResourcesRequest message is extended to pass down unmodified
+resource requests from the PodSpec.
+
```

Comment:
currently in VPA KEP there's a proposal to add UpdatePodSandboxResourcesRequest https://github.com/kubernetes/enhancements/pull/4704/files#diff-14542847beb0f0fd767db1aff1316f8569a968385e2bb89567c4cc0af1ae5942R348 which should have the pod level KubernetesResources, IMO cc @tallclair

Reply:
Again, very good point. As with https://github.com/kubernetes/enhancements/pull/4113#discussion_r1783367961 I'll update the KEP tomorrow.

Reply:
@haircommander now the proposal is updated with a reference to the in-place-pod-updates-beta (a new section on UpdatePodSandboxResourcesRequest)